feat(alert/report): Added optional CC and BCC fields for email notifi…

nsivarajan opened the pull request:
### SUMMARY
Addition of an option to provide **cc and bcc list** for email notifications along with email validation under Alerts and Reports.

### SAMPLE SCREENSHOT

<img width="382" alt="image" src="https://github.com/apache/superset/assets/117266407/7ea6c02a-ec05-4a46-a887-2d7b25ee0322">


### TESTING INSTRUCTIONS

- Open Superset in browser > Click on the "Settings" menu on top right > Click on "Alerts & Reports"
- Create a new alert/report or edit an existing alert/report. Enter the required details.
- Select "Email" as the preferred notification method.
- Enter valid/invalid email ids in the To field, CC and BCC fields and hit add/save to test the feature.

### CONTRIBUTORS
nsivarajan, episkey24

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [X] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
feat(alert/report): Added optional CC and BCC fields for email notifications

nytai commented:
@nsivarajan can you please describe why this features is useful and what problem it's solving? I'm not sure it's clear why having someone cc'd on the notification email would be any different than just adding directly to the recipients list. For BCC I can see a use case where someone would silently like to know if the report is delivered as expected, but are there any others? 


Please address the failing checks and add some unit/integration tests for this functionality

Commit:
Add test and handle build failure

nsivarajan commented:
@nytai Thanks for raising these points. In our use of Superset's alerts and reports feature, TO recipients receive actionable items, CC recipients—including leadership, hierarchical levels, and cross-functional teams—benefit from increased visibility into the communication without being directly responsible for action, while BCC recipients discreetly monitor email delivery and processes.

I've addressed the failing checks and added test cases,Could you please assist with triggering the workflow/checks again?

Commit:
handle build failure and prettify

Commit:
pytest and pre-commit validated

villebro commented on superset-frontend/src/features/alerts/AlertReportModal.tsx:
```diff
@@ -1010,6 +1014,29 @@ const AlertReportModal: FunctionComponent<AlertReportModalProps> = ({
+    const validateEmails = (emails: string): boolean => {
+      if (!emails) return true; // No emails to validate
+      return emails
+        .split(/[,;]/)
```
Should we add whitespace?
```suggestion
        .split(/[,\s;]/)
```

nsivarajan replied:
Thanks @villebro for reviewing. The Alerts and Report modal for mail notifications accepts input as either comma (",") or semicolon (";") separated values. Therefore, whitespace parsing has been avoided.


<img width="350" alt="image" src="https://github.com/apache/superset/assets/117266407/7b6311b3-d4e6-4cea-abd6-e256eda3ac65">

nsivarajan replied:
@villebro - Shall I resolve this comment ?. @kasiazjc Could you please help me here ?

villebro replied:
Thanks for clarifying - resolving the comment 👍

villebro reviewed (commented):
Code LGTM. However, given all the design work on this modal lately, it'd be nice to get @kasiazjc feedback on the new elements. One thing I believe I've seen elsewhere is CC and BCC are text/button like objects, and they expand into an input field when clicked, so as to save real estate when not needed.

nsivarajan commented:
@kasiazjc could you please assist with the review of this PR?

Commit:
avoided loop in favor of array iterations

eschutho commented:
@nsivarajan curious if you're looking for cc/bcc for each report, or if you want the same cc/bcc list for all reports. If it's the later, there's a [recent feature](https://github.com/apache/superset/pull/27851) that would allow you to add cc to all emails. We could also add a bcc there.

nsivarajan commented:
@eschutho Thanks for checking. Each report will have its own configurable CC and BCC recipients, allowing for specific recipients to be designated as needed.

 For example, the Sales dashboard will have its own CC and BCC recipients, and similarly, the Manufacturing dashboard will have its own CC and BCC recipients. This approach provides flexibility and ensures that the appropriate people are included for each report.

kasiazjc commented:
> @kasiazjc Could you please assist with the review of this PR when you have a moment? Your feedback would be greatly appreciated.

Hi, thanks for this change! In terms of design, some **general things**: 
- we don't need "optional" in the label, as already the required fields are marked with "*", so it's one or the other 
- metadata text should be under the email field, not subject 

Flow changes - agree with @villebro that we should save space and use "ghost/inline buttons", **visual specs** for these are: 
- 12px font 
- icon "mail" from ant-d; size of the icon 12px x 12px 
- spacing between the label and icon 4px
- spacing between two buttons 16px 

In terms of **behavior**: when you click on the button, it shows the field. There is no way to close it (not to overcomplicate), but if it's saved empty, when you open the modal back again it goes back to being a button

<img width="746" alt="image" src="https://github.com/apache/superset/assets/36897697/6f0c236d-ed5c-470d-9641-c1eb18fd324e">

Hope this works, let me know!

Commit:
Refactor email field UI per design feedback

nsivarajan commented:
@kasiazjc Thank you for your great suggestions and detailed review! I've incorporated the feedback and completed the changes with a commit.

Could you please assist with reviewing the PR to ensure everything aligns correctly? Your help in getting this merged would be greatly appreciated.


<img width="367" alt="Screenshot 2024-07-08 at 10 04 37 PM" src="https://github.com/apache/superset/assets/117266407/dfba0f93-5832-43ed-9308-c446d40e6456">
<img width="367" alt="Screenshot 2024-07-08 at 10 06 07 PM" src="https://github.com/apache/superset/assets/117266407/b8f47f7d-0cdf-448d-ba3b-958fd88b2fa5">
<img width="370" alt="Screenshot 2024-07-08 at 10 07 38 PM" src="https://github.com/apache/superset/assets/117266407/128a38c4-af37-49db-8b5e-074ebb06b280">
<img width="367" alt="Screenshot 2024-07-08 at 10 08 50 PM" src="https://github.com/apache/superset/assets/117266407/e267ae22-a2ec-4b8b-be35-8bbc51f8aa3a">

kasiazjc commented:
> @kasiazjc Thank you for your great suggestions and detailed review! I've incorporated the feedback and completed the changes with a commit.
> 
> Could you please assist with reviewing the PR to ensure everything aligns correctly? Your help in getting this merged would be greatly appreciated.
> 
> <img alt="Screenshot 2024-07-08 at 10 04 37 PM" width="367" src="https://private-user-images.githubusercontent.com/117266407/346650506-dfba0f93-5832-43ed-9308-c446d40e6456.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MzkxMjYsIm5iZiI6MTcyMDUzODgyNiwicGF0aCI6Ii8xMTcyNjY0MDcvMzQ2NjUwNTA2LWRmYmEwZjkzLTU4MzItNDNlZC05MzA4LWM0NDZkNDBlNjQ1Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwOVQxNTI3MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OGJlMWQ3NDYwOTU3NWFmYmEyMGJlNjJlZGJiMWE3Mzg2ODBlNmFkNzZlOGYxZmM0YzZjZjY5NzhhYzliYjYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7XAX7zKZ9PbVsxSOL1VV5f9PtdVxPo61UPOiKzascLw"> <img alt="Screenshot 2024-07-08 at 10 06 07 PM" width="367" src="https://private-user-images.githubusercontent.com/117266407/346650524-b8f47f7d-0cdf-448d-ba3b-958fd88b2fa5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MzkxMjYsIm5iZiI6MTcyMDUzODgyNiwicGF0aCI6Ii8xMTcyNjY0MDcvMzQ2NjUwNTI0LWI4ZjQ3ZjdkLTBjZGYtNDQ4ZC1iYTNiLTk1OGZkODhiMmZhNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwOVQxNTI3MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYWE0ODllZWMzMzBiMjVlMGE0ZGUwNzgzNWY0NDAyOWMzOGM2YTg2YmEyMTA3N2VlODZhZTIzMThkMzZjZTViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.48L-OtzseDFNkJAWBPKZaJGYs7JE333hEzgYZazuvSo"> <img alt="Screenshot 2024-07-08 at 10 07 38 PM" width="370" src="https://private-user-images.githubusercontent.com/117266407/346650528-128a38c4-af37-49db-8b5e-074ebb06b280.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MzkxMjYsIm5iZiI6MTcyMDUzODgyNiwicGF0aCI6Ii8xMTcyNjY0MDcvMzQ2NjUwNTI4LTEyOGEzOGM0LWFmMzctNDlkYi04YjVlLTA3NGViYjA2YjI4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwOVQxNTI3MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zM2NmODZkMDZiMTYwYjBiMGYyNjFkNjdiNGEwNmNmODg1ZmMzZWM2NjRhZjgwNGE0NjExYzg1OTQ0MmNkYTM4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2PQdnmzApeYhTCZIIfSXxjRHmQAFhKtgX64rvR6vHto"> <img alt="Screenshot 2024-07-08 at 10 08 50 PM" width="367" src="https://private-user-images.githubusercontent.com/117266407/346650531-e267ae22-a2ec-4b8b-be35-8bbc51f8aa3a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MzkxMjYsIm5iZiI6MTcyMDUzODgyNiwicGF0aCI6Ii8xMTcyNjY0MDcvMzQ2NjUwNTMxLWUyNjdhZTIyLWEyZWMtNGI4Yi1iZTM1LThiYmM1MWY4YWEzYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwOVQxNTI3MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZGE2YTczNzY1NGI3ODczNjQxM2IyZWFmMmIzZWRlZDZmZmRjYTkzMzliMDY2YjgyZjMxNGIzYTQxOTZiMzhiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.sxWQ25Fh_g0NaS9HIMND32OX7n6sc3dqscAYU3_u-_w">

Thank you for the changes, this looks good! I have two small things: 
- I think envelope icon might be bigger than 12px, can you make it smaller? Also align both text and icon to the center (should be looking similar to the "button" below with "+ add new notification method" 
- when you add one field, the button is not aligned to the left "add bcc..." and there is some space left, can it be in one line with all of the other fields etc aligned to the left? 
<img width="256" alt="image" src="https://github.com/apache/superset/assets/36897697/9349b776-5e90-45f9-8ebf-a24f671863e9">

<img width="256" alt="image" src="https://github.com/apache/superset/assets/36897697/0ba14f6a-acc0-465b-b838-ee6b957179e8">

Commit:
Refine icon size and alignment in notification settings

nsivarajan commented:
@kasiazjc 

Thank you for your feedback! I've incorporated the requested adjustments into the PR.

Could you please review the changes once more to confirm they meet the requirements? Additionally, I would appreciate your assistance in merging this PR.


<img width="516" alt="Screenshot 2024-07-10 at 2 01 32 PM" src="https://github.com/apache/superset/assets/117266407/28802033-3b92-4123-86dd-4a7b516f6bbf">
<img width="377" alt="Screenshot 2024-07-10 at 4 30 45 PM" src="https://github.com/apache/superset/assets/117266407/aa4aca16-8e72-4f3a-9a98-3243a0a1bf60">
<img width="374" alt="Screenshot 2024-07-10 at 4 33 03 PM" src="https://github.com/apache/superset/assets/117266407/46079bc8-a953-4b95-b399-962b8a841403">

kgabryje commented on superset-frontend/src/features/alerts/components/NotificationMethod.tsx:
```diff
@@ -36,6 +36,12 @@ const StyledNotificationMethod = styled.div`
     }
+
+    .helper {
+      margin-top: 5px;
```
Can we use theme variables for all spacings and font sizes?

nsivarajan replied:
@kgabryje Thank you for the review! I've refactored styles to use theme variables for spacings and font sizes.

nsivarajan replied:
@kgabryje Could you please review and resolve the conversation?

kgabryje replied:
Sorry for the delay! As a rule we try to avoid fractions of grid units like `gridUnit * 2.5` or `gridUnit * 0.75`. Could you please round those values to whole grid units?

nsivarajan replied:
Before making adjustments, I'd like confirmation on aligning with our guideline to avoid fractional grid units. Based on our 4px gridUnit definition, here are the proposed changes:

```
- margin-bottom: ${theme.gridUnit * 2.5}px;
+ margin-bottom: ${theme.gridUnit * 2 + 2}px;

- margin-left: ${theme.gridUnit * 2.5}px;
- padding-top: ${theme.gridUnit * 0.75}px;
+ margin-left: ${theme.gridUnit * 2 + 2}px;
+ padding-top: ${theme.gridUnit - 1}px;
```

Please review and confirm so I can proceed accordingly.

kgabryje replied:
Not exactly - we use the grid units to enforce consistent spacing between elements and to maintain that consistent spacing in case the grid unit is changed (for example as a result of a user customising the theme). So in other words, we should just use full multiples of grid unit.
Of course that takes away some flexibility, as you can't set a margin of 5, 6 or 7 pixels. You can just pick the closest grid unit (the closest multiple of 4) that looks good

kgabryje replied:
Side note - you might notice that there are some components that use fractions of grid units. However, those are old components created before we started enforcing that rule and those haven't been refactored yet. In new code we should stick to the full grid units

nsivarajan replied:
@kgabryje Thanks for detailed explanation.

 I have done below changes and pushed commit. Please assist with triggering the CI and resolving comment, and bringing up ephemeral env as needed.

```
-      margin-bottom: ${theme.gridUnit * 2.5}px;
+      margin-bottom: ${theme.gridUnit * 2}px;
-        margin-left: ${theme.gridUnit * 2.5}px;
-        padding-top: ${theme.gridUnit * 0.75}px;
+        margin-left: ${theme.gridUnit * 2}px;
+        padding-top: ${theme.gridUnit}px;

```

kgabryje replied:
Thank you! Spotted 1 remaining fraction, and we can just use 0px instead of theme.gridUnit * 0 🙂

kgabryje commented on superset-frontend/src/features/alerts/components/NotificationMethod.tsx:
```diff
@@ -178,57 +256,134 @@ export const NotificationMethod: FunctionComponent<NotificationMethodProps> = ({
+              <div
+                style={{
+                  color: theme.colors.error.base,
+                  fontSize: theme.gridUnit * 3,
```
We’ve got specific theme variables for font sizes - check out typography part of superset theme

nsivarajan replied:
@kgabryje Could you please review and resolve the conversation?

Commit:
Refactored styles to use theme

villebro commented:
@nsivarajan I've started CI and I'll start an eph env once the Docker image is ready so we can poke around on the modal.

rusackas commented:
restarted some failing tests... hoping they're just flaky 🤞

nsivarajan commented:
Thanks @rusackas,

The Python-Integration tests for test-mysql (pull_request) and test-postgres (current) (pull_request) are failing, although both passed successfully on my local setup. A retrigger should resolve this.

Appreciate your assistance.

villebro commented:
/testenv up FLAG_ALERT_REPORTS=True

eschutho commented:
FEATURE_ALERT_REPORTS=True /testenv up

eschutho commented:
/testenv up FEATURE_ALERT_REPORTS=true

nsivarajan commented:
@villebro @eschutho Thank you for spinning up the ephemeral environment. Could you please review and assist with merging this PR? Your support in moving this forward would be greatly appreciated.

Commit:
Avoid fractions of grid units

kgabryje commented on superset-frontend/src/features/alerts/components/NotificationMethod.tsx:
```diff
@@ -24,34 +24,77 @@ import { NotificationMethodOption, NotificationSetting } from '../types';
 const StyledNotificationMethod = styled.div`
-  margin-bottom: 10px;
+  ${({ theme }) => `
+    margin-bottom: ${theme.gridUnit * 2.5}px;
```
```suggestion
    margin-bottom: ${theme.gridUnit * 2}px;
```
(or `* 3`)

nsivarajan replied:
Oh yes, updated now with ```margin-bottom: ${theme.gridUnit * 3}px;```

kgabryje commented on superset-frontend/src/features/alerts/components/NotificationMethod.tsx:
```diff
@@ -24,34 +24,77 @@ import { NotificationMethodOption, NotificationSetting } from '../types';
+      margin-bottom: ${theme.gridUnit * 2}px;
+
+      > div {
+        margin: ${theme.gridUnit * 0}px;
```
If it's 0 it's fine to just put `0px` since that won't really depend on gridUnit's value 😄

nsivarajan replied:
Thanks! corrected to ```margin: 0px;```

nsivarajan replied:
Please could you suggest, can I change `role` with `data-testid` . 

like below


```
<span
    data-testid="delbutton"
    tabIndex={0}
    className="delete-button"
    onClick={() => onRemove(index)}
>
```

```
<span
    data-testid="ccbutton"
    tabIndex={0}
    className="delete-button"
    onClick={() => onRemove(index)}
>
```


              
and test case for specific delete button as 

```
const deleteButton = screen.getByTestId('delbutton');
userEvent.click(deleteButton);

```

kgabryje commented:
/testenv up

kgabryje commented:
/testenv up FEATURE_ALERT_REPORTS=true

Commit:
Resolve conflicts and improve styling

nsivarajan commented on superset-frontend/src/features/alerts/components/NotificationMethod.tsx:
```diff
@@ -24,34 +24,77 @@ import { NotificationMethodOption, NotificationSetting } from '../types';
+
+    .ghost-button:first-child[style*='none'] + .ghost-button {
+      margin-left: ${
+        theme.gridUnit * 0
```
Thanks! corrected to ```margin-left: 0px;```

kgabryje replied:
Thanks!

kgabryje reviewed (approved):
LGTM, thanks for addressing the feedback!

Commit:
Handle failed test case for delete button

kgabryje commented on superset-frontend/src/features/alerts/components/NotificationMethod.tsx:
```diff
@@ -437,7 +437,7 @@ export const NotificationMethod: FunctionComponent<NotificationMethodProps> = ({
             {index !== 0 && !!onRemove ? (
               <span
-                role="button"
+                role="delbutton"
```
`role` attribute should only use [aria roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) since the application's accessibility depends on it.

nsivarajan replied:
Please could you suggest , can I change `role` with `data-testid`

like below

```
               <span
                  className="ghost-button"
                  data-testid="ccbutton"
                  tabIndex={0}
                  onClick={() => setCcVisible(true)}
                  style={{ display: ccVisible ? 'none' : 'inline-flex' }}
                >
```

```
              <span
                data-testid="delbutton"
                tabIndex={0}
                className="delete-button"
                onClick={() => onRemove(index)}
              >

```

```
                <span
                  className="ghost-button"
                  data-testid="bccbutton"
                  tabIndex={0}
                  onClick={() => setBccVisible(true)}
                  style={{ display: bccVisible ? 'none' : 'inline-flex' }}
                >
```

and relevant test case to evaluate real delete is as below

```
const deleteButton = screen.getByTestId('delbutton');
userEvent.click(deleteButton);

```

nsivarajan replied:
Thanks @kgabryje !  Updated test case and validated to be success.

```
npm run test -- --silent src/features/alerts/components/NotificationMethod.test.tsx

 PASS  src/features/alerts/components/NotificationMethod.test.tsx

Test Suites: 1 passed, 1 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        4.149 s, estimated 7 s

```

nsivarajan replied:
@kgabryje 

The "Python-Integration / test-mysql (pull_request)" job is failing intermittently and appears to be flaky. A retrigger or reattempt should succeed.

kgabryje replied:
yeah its definitely flaky... I re-run it, hopefully it passes now

nsivarajan replied:
Thanks again, @kgabryje! All tests are now successful. Please assist with merging this PR.

nsivarajan replied:
Thanks again, @kgabryje! All tests are now successful. Please assist with merging this PR.

Commit:
Handle failed test case as per suggestion

kgabryje commented:
Merged, thanks for the contribution @nsivarajan!
